File: src/models/cleaner.py
"""智能清理引擎 — Electron 缓存猎手、大文件扫描、临时文件清理。"""

import os
import shutil
import logging
from dataclasses import dataclass, field
from pathlib import Path

import psutil

from src.utils.constants import (
    ELECTRON_CACHE_DIRS,
    TEMP_EXTENSIONS,
    LARGE_FILE_THRESHOLD_BYTES,
)
from src.utils.winapi import empty_recycle_bin

logger = logging.getLogger(__name__)


@dataclass
class CleanItem:
    """单个可清理项。"""
    path: str
    size: int
    category: str  # "electron" | "temp" | "large_file"
    description: str = ""
    selected: bool = True


@dataclass
class ScanResult:
    """扫描结果汇总。"""
    items: list[CleanItem] = field(default_factory=list)

    @property
    def total_size(self) -> int:
        return sum(i.size for i in self.items if i.selected)

    @property
    def count(self) -> int:
        return sum(1 for i in self.items if i.selected)


# ── Electron 缓存猎手 ─────────────────────────────────────────────


def _has_electron_signature(directory: str) -> bool:
    """检查目录是否具有 Electron 应用的特征结构。"""
    markers = {"Cache", "GPUCache"}
    try:
        children = {e.name for e in os.scandir(directory) if e.is_dir()}
    except OSError:
        return False

    if not markers.issubset(children):
        return False

    # 在前两级目录中查找 .asar 文件
    for root, dirs, files in os.walk(directory):
        if any(f.endswith(".asar") for f in files):
            return True
        depth = root.replace(directory, "").count(os.sep)
        if depth >= 2:
            break
    return False


def _dir_size(path: str) -> int:
    """迭代计算目录总大小（栈模拟，避免栈溢出）。"""
    total = 0
    stack = [path]
    while stack:
        current = stack.pop()
        try:
            for entry in os.scandir(current):
                if entry.is_file(follow_symlinks=False):
                    total += entry.stat(follow_symlinks=False).st_size
                elif entry.is_dir(follow_symlinks=False):
                    stack.append(entry.path)
        except OSError:
            pass
    return total


def scan_electron_caches() -> list[CleanItem]:
    """通过签名匹配扫描 AppData 中的 Electron 应用缓存。"""
    items: list[CleanItem] = []
    roots = []
    for var in ("APPDATA", "LOCALAPPDATA"):
        val = os.environ.get(var)
        if val and os.path.isdir(val):
            roots.append(val)

    for root in roots:
        try:
            entries = list(os.scandir(root))
        except OSError:
            continue
        for entry in entries:
            if not entry.is_dir():
                continue
            # 检查顶层及下一级子目录
            paths_to_check = [entry.path]
            try:
                paths_to_check.extend(
                    sub.path for sub in os.scandir(entry.path)
                    if sub.is_dir()
                )
            except OSError:
                pass

            for check_path in paths_to_check:
                if not _has_electron_signature(check_path):
                    continue
                for cache_name in ELECTRON_CACHE_DIRS:
                    cache_path = os.path.join(check_path, cache_name)
                    if os.path.isdir(cache_path):
                        size = _dir_size(cache_path)
                        if size > 0:
                            items.append(CleanItem(
                                path=cache_path,
                                size=size,
                                category="electron",
                                description=f"Electron: {Path(check_path).name}",
                            ))
    return items


# ── 临时文件扫描 ───────────────────────────────────────────────────


def scan_temp_files() -> list[CleanItem]:
    """扫描系统和用户临时目录。"""
    items: list[CleanItem] = []
    temp_dirs = set()
    for var in ("TEMP", "TMP"):
        val = os.environ.get(var)
        if val and os.path.isdir(val):
            temp_dirs.add(val)
    win_temp = os.path.join(os.environ.get("SYSTEMROOT", r"C:\Windows"), "Temp")
    if os.path.isdir(win_temp):
        temp_dirs.add(win_temp)

    for temp_dir in temp_dirs:
        try:
            for entry in os.scandir(temp_dir):
                try:
                    if entry.is_file(follow_symlinks=False):
                        size = entry.stat(follow_symlinks=False).st_size
                        items.append(CleanItem(
                            path=entry.path,
                            size=size,
                            category="temp",
                            description=f"临时文件: {entry.name}",
                        ))
                    elif entry.is_dir(follow_symlinks=False):
                        size = _dir_size(entry.path)
                        items.append(CleanItem(
                            path=entry.path,
                            size=size,
                            category="temp",
                            description=f"临时目录: {entry.name}",
                        ))
                except OSError:
                    continue
        except OSError:
            continue
    return items


# ── 大文件扫描 ─────────────────────────────────────────────────────


def _scan_large_files_mft(
    drive_letter: str,
    threshold: int,
    max_results: int,
) -> list[CleanItem] | None:
    """尝试通过 MFT 快速扫描大文件（需要管理员权限）。失败返回 None。"""
    try:
        from src.models.mft_scanner import scan_mft_large_files
        entries = scan_mft_large_files(
            drive_letter=drive_letter,
            min_size_bytes=threshold,
            max_results=max_results,
        )
        if not entries:
            return None
        items = []
        for e in entries:
            ext = os.path.splitext(e.path)[1].lower()
            items.append(CleanItem(
                path=e.path,
                size=e.size,
                category="large_file",
                description=e.path,
                selected=ext in TEMP_EXTENSIONS,
            ))
        return items
    except Exception:
        logger.debug("MFT 扫描不可用，回退到 os.walk", exc_info=True)
        return None


def _scan_large_files_walk(
    root: str,
    threshold: int,
    max_results: int,
) -> list[CleanItem]:
    """通过 os.walk 遍历扫描大文件（兜底方案）。"""
    items: list[CleanItem] = []
    skip_prefixes = (
        os.path.join(root, "Windows"),
        os.path.join(root, "Program Files"),
        os.path.join(root, "Program Files (x86)"),
        os.path.join(root, "$"),
    )

    for dirpath, dirnames, filenames in os.walk(root, topdown=True):
        if any(dirpath.startswith(p) for p in skip_prefixes):
            dirnames.clear()
            continue
        dirnames[:] = [
            d for d in dirnames
            if not d.startswith("$") and not d.startswith(".")
        ]
        for fname in filenames:
            fpath = os.path.join(dirpath, fname)
            try:
                size = os.path.getsize(fpath)
            except OSError:
                continue
            if size >= threshold:
                ext = os.path.splitext(fname)[1].lower()
                items.append(CleanItem(
                    path=fpath,
                    size=size,
                    category="large_file",
                    description=fpath,
                    selected=ext in TEMP_EXTENSIONS,
                ))
                if len(items) >= max_results:
                    return items
    return items


def scan_large_files(
    root: str = "C:\\",
    threshold: int = LARGE_FILE_THRESHOLD_BYTES,
    max_results: int = 200,
) -> list[CleanItem]:
    """扫描大文件。优先使用 MFT 快速扫描，失败时回退到 os.walk。"""
    drive_letter = root.rstrip("\\").rstrip(":")
    if len(drive_letter) == 1 and drive_letter.isalpha():
        result = _scan_large_files_mft(drive_letter, threshold, max_results)
        if result is not None:
            return result
    return _scan_large_files_walk(root, threshold, max_results)


# ── 清理执行 ───────────────────────────────────────────────────────


def _is_path_locked(path: str) -> bool:
    """检查是否有运行中的进程锁定了该路径。"""
    path_lower = os.path.normpath(path).lower()
    for proc in psutil.process_iter(["pid", "name"]):
        try:
            for f in proc.open_files():
                if os.path.normpath(f.path).lower().startswith(path_lower):
                    return True
        except (psutil.AccessDenied, psutil.NoSuchProcess):
            continue
    return False


def _safe_remove_file(path: str) -> int:
    """安全删除单个文件，返回释放的字节数。失败返回 0。"""
    try:
        size = os.path.getsize(path)
        os.remove(path)
        return size
    except (PermissionError, OSError):
        return 0


def _safe_remove_tree(path: str) -> int:
    """安全删除目录树，跳过被占用的文件，返回实际释放的字节数。"""
    freed = 0
    for root, dirs, files in os.walk(path, topdown=False):
        for name in files:
            freed += _safe_remove_file(os.path.join(root, name))
        for name in dirs:
            try:
                os.rmdir(os.path.join(root, name))
            except OSError:
                pass
    try:
        os.rmdir(path)
    except OSError:
        pass
    return freed


def execute_clean(items: list[CleanItem]) -> tuple[int, int]:
    """删除已选中的项目。返回 (已清理字节数, 失败数量)。

    对被占用或无权限的文件静默跳过，不中断整体清理流程。
    """
    cleaned = 0
    failed = 0
    for item in items:
        if not item.selected:
            continue
        try:
            if item.category == "recycle_bin":
                if empty_recycle_bin():
                    cleaned += item.size
                else:
                    failed += 1
            elif os.path.isdir(item.path):
                freed = _safe_remove_tree(item.path)
                cleaned += freed
                if freed == 0:
                    failed += 1
            elif os.path.isfile(item.path):
                freed = _safe_remove_file(item.path)
                if freed > 0:
                    cleaned += freed
                else:
                    failed += 1
            else:
                # 文件已不存在，视为已清理
                cleaned += item.size
        except Exception:
            logger.warning("清理失败: %s", item.path, exc_info=True)
            failed += 1
    return cleaned, failed

File: Models/Cleaner.cs
using PrismCore.Helpers;

namespace PrismCore.Models;

/// <summary>智能清理引擎（对照 cleaner.py）。</summary>
public static class Cleaner
{
    public class CleanItem(string path, long size, string category, string description, bool selected = true)
    {
        public string Path { get; } = path;
        public long Size { get; } = size;
        public string Category { get; } = category;
        public string Description { get; } = description;
        public bool Selected { get; set; } = selected;
        public string SizeText => SystemInfo.FormatBytes((ulong)Size);
    }

    public record ScanResult(List<CleanItem> Items)
    {
        public long TotalSize => Items.Where(i => i.Selected).Sum(i => i.Size);
        public int Count => Items.Count(i => i.Selected);
    }

    // ── Electron 缓存扫描 ──

    public static List<CleanItem> ScanElectronCaches()
    {
        var items = new List<CleanItem>();
        var roots = new List<string>();
        foreach (var v in new[] { "APPDATA", "LOCALAPPDATA" })
        {
            var val = Environment.GetEnvironmentVariable(v);
            if (val != null && Directory.Exists(val)) roots.Add(val);
        }

        foreach (var root in roots)
        {
            try
            {
                foreach (var entry in Directory.GetDirectories(root))
                {
                    var toCheck = new List<string> { entry };
                    try { toCheck.AddRange(Directory.GetDirectories(entry)); } catch { }

                    foreach (var dir in toCheck)
                    {
                        if (!HasElectronSignature(dir)) continue;
                        foreach (var cacheName in Constants.ElectronCacheDirs)
                        {
                            var cachePath = Path.Combine(dir, cacheName);
                            if (!Directory.Exists(cachePath)) continue;
                            var size = DirSize(cachePath);
                            if (size > 0)
                                items.Add(new(cachePath, size, "electron",
                                    $"Electron: {Path.GetFileName(dir)}"));
                        }
                    }
                }
            }
            catch { }
        }
        return items;
    }

    private static bool HasElectronSignature(string dir)
    {
        try
        {
            var children = Directory.GetDirectories(dir).Select(Path.GetFileName).ToHashSet();
            if (!children.Contains("Cache") || !children.Contains("GPUCache")) return false;
            // 查找前两级 .asar 文件
            foreach (var f in Directory.EnumerateFiles(dir, "*.asar", SearchOption.AllDirectories))
                return true;
        }
        catch { }
        return false;
    }

    // ── 临时文件扫描 ──

    public static List<CleanItem> ScanTempFiles()
    {
        var items = new List<CleanItem>();
        var tempDirs = new HashSet<string>();
        foreach (var v in new[] { "TEMP", "TMP" })
        {
            var val = Environment.GetEnvironmentVariable(v);
            if (val != null && Directory.Exists(val)) tempDirs.Add(val);
        }
        var winTemp = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.Windows), "Temp");
        if (Directory.Exists(winTemp)) tempDirs.Add(winTemp);

        foreach (var tempDir in tempDirs)
        {
            try
            {
                foreach (var entry in new DirectoryInfo(tempDir).EnumerateFileSystemInfos())
                {
                    try
                    {
                        if (entry is FileInfo fi)
                            items.Add(new(fi.FullName, fi.Length, "temp", $"临时文件: {fi.Name}"));
                        else if (entry is DirectoryInfo di)
                            items.Add(new(di.FullName, DirSize(di.FullName), "temp", $"临时目录: {di.Name}"));
                    }
                    catch { }
                }
            }
            catch { }
        }
        return items;
    }

    // ── 大文件扫描 ──

    public static List<CleanItem> ScanLargeFiles(
        string root = @"C:\", long threshold = 0, int maxResults = 200)
    {
        if (threshold == 0) threshold = Constants.LargeFileThresholdBytes;
        var letter = root.TrimEnd('\\').TrimEnd(':');
        if (letter.Length == 1 && char.IsLetter(letter[0]))
        {
            var mftResult = ScanLargeFilesMft(letter, threshold, maxResults);
            if (mftResult != null) return mftResult;
        }
        return ScanLargeFilesWalk(root, threshold, maxResults);
    }

    private static List<CleanItem>? ScanLargeFilesMft(string letter, long threshold, int max)
    {
        try
        {
            var entries = MftScanner.ScanLargeFiles(letter, threshold, max);
            if (entries.Count == 0) return null;
            return entries.Select(e =>
            {
                var ext = Path.GetExtension(e.Path).ToLowerInvariant();
                return new CleanItem(e.Path, e.Size, "large_file", e.Path,
                    Constants.TempExtensions.Contains(ext));
            }).ToList();
        }
        catch { return null; }
    }

    private static List<CleanItem> ScanLargeFilesWalk(string root, long threshold, int max)
    {
        var items = new List<CleanItem>();
        var skipPrefixes = new[]
        {
            Path.Combine(root, "Windows"), Path.Combine(root, "Program Files"),
            Path.Combine(root, "Program Files (x86)"), Path.Combine(root, "$")
        };

        var stack = new Stack<string>();
        stack.Push(root);
        while (stack.Count > 0 && items.Count < max)
        {
            var dir = stack.Pop();
            if (skipPrefixes.Any(dir.StartsWith)) continue;
            try
            {
                foreach (var d in Directory.GetDirectories(dir))
                {
                    var name = Path.GetFileName(d);
                    if (!name.StartsWith('$') && !name.StartsWith('.'))
                        stack.Push(d);
                }
                foreach (var f in Directory.GetFiles(dir))
                {
                    try
                    {
                        var fi = new FileInfo(f);
                        if (fi.Length >= threshold)
                        {
                            var ext = fi.Extension.ToLowerInvariant();
                            items.Add(new(f, fi.Length, "large_file", f,
                                Constants.TempExtensions.Contains(ext)));
                            if (items.Count >= max) break;
                        }
                    }
                    catch { }
                }
            }
            catch { }
        }
        return items;
    }

    // ── 清理执行 ──

    public static (long Cleaned, int Failed) ExecuteClean(List<CleanItem> items)
    {
        long cleaned = 0;
        int failed = 0;
        foreach (var item in items.Where(i => i.Selected))
        {
            try
            {
                if (item.Category == "recycle_bin")
                {
                    if (NativeApi.SHEmptyRecycleBinW(0, null, NativeApi.SHERB_NO_UI) == 0)
                        cleaned += item.Size;
                    else failed++;
                }
                else if (Directory.Exists(item.Path))
                {
                    var freed = SafeRemoveTree(item.Path);
                    cleaned += freed;
                    if (freed == 0) failed++;
                }
                else if (File.Exists(item.Path))
                {
                    var size = new FileInfo(item.Path).Length;
                    File.Delete(item.Path);
                    cleaned += size;
                }
                else cleaned += item.Size; // 已不存在
            }
            catch { failed++; }
        }
        return (cleaned, failed);
    }

    private static long SafeRemoveTree(string path)
    {
        long freed = 0;
        try
        {
            foreach (var f in Directory.EnumerateFiles(path, "*", SearchOption.AllDirectories))
            {
                try { var len = new FileInfo(f).Length; File.Delete(f); freed += len; } catch { }
            }
            try { Directory.Delete(path, true); } catch { }
        }
        catch { }
        return freed;
    }

    private static long DirSize(string path)
    {
        long total = 0;
        var stack = new Stack<string>();
        stack.Push(path);
        while (stack.Count > 0)
        {
            var dir = stack.Pop();
            try
            {
                foreach (var f in Directory.GetFiles(dir))
                    try { total += new FileInfo(f).Length; } catch { }
                foreach (var d in Directory.GetDirectories(dir))
                    stack.Push(d);
            }
            catch { }
        }
        return total;
    }
}
